### DISCUSSION TITLE
Router middleware composable

### DISCUSSION THREAD

**Discussion by atinux** [nuxt/nuxt#17908] [Announcements]
In Nuxt 2 we do have the [`middleware/` directory](https://nuxtjs.org/docs/directory-structure/middleware).

We could have the same with Nuxt 3 but instead of using the `middleware` property, setup a `useMiddleware()` composable that could be use in any Vue components.

```ts
// middleware/auth.ts
export default defineMiddleware((from, to) => {
  const session = useSession()

  if (!hasSession()) {
    return { redirect: '/login' }
  }
})
```

Then in the application, we can use `useMiddleware(name)` in the pages and components:

```vue
<script setup>
useMiddleware('auth')
</script>
```

## First attempt

I made an initital POC but it seems that `useMiddleware` is acting more as a `beforeRouteLeave` hook since it register the guard only once the component is already instanciated.

https://stackblitz.com/edit/nuxt-middleware-attempt?file=composables%2FuseMiddleware.ts

1. Go to `/secret` page (you are not redirected)
2. Click to `/more` (no middleware registered but got redirected to /)

Let's see with @posva how we can find a solution for it mixing `vue-router` and `<Suspense>`.

**Comment by pi0** [comment]
Apart from implementation, some toughts:

- If the goal is to guard navigation and redirect, maybe rename middleware to Navigation Guards (`useGuard`)?
- It would be nice if we assume guards can be async as well to allow async logic
- Guards/Middleware should be usable within `app.vue` as well (without pages) for single page auth
- We might also support conditional guards by providing `useMiddleware`/`useGuard` instead of definer:

```js
export const useAuth = (adminRequired: false) => useMiddleware((from, to) => {
  if (!hasSession() || (!useSession().isAdmin && adminRequired)) {
    return { redirect: '/login' }
  }
})
```


```vue
<script setup>
await useAuth (true /* should be authenticated and admin */)
</script>
```

**Reply by atinux** [reply]
I wanted to keep middleware since it may offer other features such as simply logging or metrics, but I guess they can do a plugin for that 🧐

**Reply by psycongaroo** [reply]
Please keep middleware! @Atinux @pi0

**Reply by pi0** [reply]
> such as simply logging or metrics, but I guess they can do a plugin for that

Indeed. We can make use of non-blocking router hooks for logging 

BTW one more reason to avoid middleware, is that with Nuxt3 we introduced server/middleware and by practice, it was SO easy that users confusing server middleware and nuxt middleware that are completely different concepts.

**Reply by acidjazz** [reply]
Maybe an example using `server/middleware` would help us understand what is better practice moving forward?

**Comment by acidjazz** [comment]
Just for input I've been working on a simple middleware composable based on @Atinux 's suggestion with an Auth plugin I'm working on, working well so far:

`composables/useAuthMiddleware.ts`

```ts
import { useNuxtApp } from '#app'
import { useRouter } from 'vue-router'
import { watchEffect } from '@vue/runtime-core'

export const useAuthMiddleware = async () => {
  const { $auth } = useNuxtApp()
  const router = useRouter()
  if ($auth.loggedIn.value === false) {
    await router.push({path: $auth.config.redirect.logout})
  }

  watchEffect(async () => {
    if ($auth.loggedIn.value === false)
      await router.push({path: $auth.config.redirect.logout})
  })
}
```

**Reply by atinux** [reply]
How well does it work with SSR?

**Reply by acidjazz** [reply]
@Atinux so far so good!

**Reply by ragokan** [reply]
This looks pretty much interesting, do you use it in parent etc? Can you show where did you use @acidjazz?

**Reply by danielroe** [reply]
We now have built-in Nuxt route middleware:  http://v3.nuxtjs.org/docs/directory-structure/middleware

**Reply by acidjazz** [reply]
> This looks pretty much interesting, do you use it in parent etc? Can you show where did you use @acidjazz?

I would refer to the official middleware that is now live:

https://v3.nuxtjs.org/docs/directory-structure/middleware

**Comment by psycongaroo** [comment]
A great idea is like Nuxt2 you can configure globally middleware inside nuxt.config.ts :)

**Comment by thomasowow** [comment]
Thanks for the inspiration. With some help of https://github.com/nuxt/framework/discussions/1722#discussioncomment-1649805 I came up with the following solution that works on both client and SSR:
```ts
export const AUTH_MIDDLEWARE_GUEST_REDIRECT = '/';

export const AUTH_MIDDLEWARE_AUTH_REDIRECT = '/auth/login';

export const useAuthMiddleware = async(guest = false) => {
    const app = useNuxtApp();

    const user = useAuthUser(); // Or however you would like to get authenticated user.

    const redirect = async(path: string) => {
        if (!!app.ssrContext) {
            app.ssrContext.res.writeHead(302, {
                Location: path,
            });

            return app.ssrContext.res.end();
        }

        return await useRouter().push(path);
    };

    if (guest && !!user.value) {
        return redirect(AUTH_MIDDLEWARE_GUEST_REDIRECT);
    } else if (!guest && !user.value) {
        return redirect(AUTH_MIDDLEWARE_AUTH_REDIRECT);
    }
};
```
Page component:
```html
<script lang="ts" setup>
await useAuthMiddleware();
<script>
```

**Reply by beocaca** [reply]
> Thanks for the inspiration. With some help of [#1722 (reply in thread)](https://github.com/nuxt/framework/discussions/1722#discussioncomment-1649805) I came up with the following solution that works on both client and SSR:
> 
> ```ts
> export const AUTH_MIDDLEWARE_GUEST_REDIRECT = '/';
> 
> export const AUTH_MIDDLEWARE_AUTH_REDIRECT = '/auth/login';
> 
> export const useAuthMiddleware = async(guest = false) => {
>     const app = useNuxtApp();
> 
>     const user = useAuthUser(); // Or however you would like to get authenticated user.
> 
>     const redirect = async(path: string) => {
>         if (!!app.ssrContext) {
>             app.ssrContext.res.writeHead(302, {
>                 Location: path,
>             });
> 
>             return app.ssrContext.res.end();
>         }
> 
>         return await useRouter().push(path);
>     };
> 
>     if (guest && !!user.value) {
>         return redirect(AUTH_MIDDLEWARE_GUEST_REDIRECT);
>     } else if (!guest && !user.value) {
>         return redirect(AUTH_MIDDLEWARE_AUTH_REDIRECT);
>     }
> };
> ```
> 
> Page component:
> 
> ```
> <script lang="ts" setup>
> await useAuthMiddleware();
> <script>
> ```

if you set useAuthUser() reactive , you can go back to previous page 
fail at router logic
ex : 
login page > login success > redirect to "/" > go back history >  fail

**Reply by psycongaroo** [reply]
@namcodes we just needs guards/middleware for client/ssr router.
I think is a bad concept do it with composables.

@Atinux @pi0 @danielroe

**Comment by acidjazz** [comment]
As I've been working in nuxt3 with pages that require auth and pages that don't, I've found that by just making sure my API calls that require auth respond with a proper `401` error - I have a wrapper around fetch in my api/auth library that treats this accordingly:

```ts
    if (error.response?.status === 401)
      return await this.invalidate()
```
This just simply redirects to the logout redirect and does some cookie/session some cleanup.

With this methodology I see no reason for auth middleware anymore - and it helps me test and make sure my endpoints have proper policies.

The template I use with this methodology is up to date and found [here](https://github.com/fumeapp/laranuxt) - the code i'm referring to is found [here](https://github.com/fumeapp/laranuxt/blob/main/client/lib/api.ts#L156)
